pandas/core/frame.py

```diff
@@ -9787,7 +9787,9 @@ def _join_compat(
                     "Suffixes not supported when joining multiple DataFrames"
                 )
 
-            frames = [self] + list(other)
+            frames = [cast("DataFrame | Series", self)] + list(
+                cast("Iterable[DataFrame | Series]", other)
```

Comment:
Since at this point you know that `other` is a list, we don't need the `list(other)` part.  Without typing, you could do `frames = [self] + other` .  It may be that you have to cast `self` because of the `+` operator, which typing would reject if concatenating a `List[DataFrame]` and `List[DataFrame | Series]`